src/main/scala/io/iohk/ethereum/consensus/blocks/BlockGenerator.scala

```diff
@@ -0,0 +1,62 @@
+package io.iohk.ethereum.consensus.blocks
+
+import akka.util.ByteString
+import io.iohk.ethereum.domain.{Address, Block, SignedTransaction}
+import io.iohk.ethereum.ledger.BlockPreparationError
+
+/**
+ * We use a `BlockGenerator` to create the next block.
+ * In a PoW setting, this is what a miner typically does.
+ * In general, a [[BlockGenerator]] depends on and is provided by the
+ * [[io.iohk.ethereum.consensus.Consensus Consensus]].
+ *
+ * @note This is generally a stateful object.
+ *
+ * @see [[io.iohk.ethereum.consensus.Consensus#blockGenerator Consensus#blockGenerator()]],
+ *      [[io.iohk.ethereum.ledger.BlockPreparator BlockPreparator]]
+ *
+ */
+trait BlockGenerator {
+  /**
+   * The type of consensus-specific data used in the block generation process.
+   * For example, under [[io.iohk.ethereum.consensus.ethash.EthashConsensus EthashConsensus]],
+   * this represents the [[io.iohk.ethereum.network.p2p.messages.PV62.BlockBody#uncleNodesList ommers]].
+   */
+  type X
+
+  /** An empty `X` */
+  def emptyX: X
+
+  // FIXME This is currently used only for Ethash
```

Comment:
why don't we put it in ethash specific generator then?

Reply:
Too much refactoring :) 
I can prepare a patch and even save us from one more [`FIXME`](https://github.com/input-output-hk/mantis/blob/feature/consensus_as_a_service_EC458/src/main/scala/io/iohk/ethereum/jsonrpc/EthService.scala#L500-L513).

The good news is that all the affected tests already expect the *correct* `BlockGenerator` subclass, that is `EthashBlockGenerator`. For example:

* [in EthServiceSpec](https://github.com/input-output-hk/mantis/blob/feature/consensus_as_a_service_EC458/src/test/scala/io/iohk/ethereum/jsonrpc/EthServiceSpec.scala#L396) we use this [type](https://github.com/input-output-hk/mantis/blob/feature/consensus_as_a_service_EC458/src/test/scala/io/iohk/ethereum/jsonrpc/EthServiceSpec.scala#L860).

Reply:
OK, I did prepare the [patch](https://github.com/input-output-hk/mantis/pull/410/commits/375064acb3dc4d89441dde359941dea1f88bb87d).